Add blog post about persistent storage improvements in 0.12.0 (#78)
strimzi/strimzi.github.io: _posts/2019-07-08-persistent-storage-improvements.md

---
layout: post
title:  "Persistent storage improvements"

> date: 2019-08-07

Reviewer: isn't 07-08 ? It should be July 8th not August 7th

author: jakub_scholz
---

> Few months ago I wrote [a blog post](https://strimzi.io/2019/02/28/resizing-persistent-volumes.html) about how you can manually increase the size of your persistent volumes used for Kafka or Zookeeper storage.

Reviewer 2: ```suggestion
A few months ago I wrote [a blog post](https://strimzi.io/2019/02/28/resizing-persistent-volumes.html) about how you can manually increase the size of the persistent volumes you use for Kafka or Zookeeper storage.
```

> I promised there that one day it will be supported directly in Strimzi.

Reviewer 2: ```suggestion
I promised that one day it would be supported directly in Strimzi.
```

> And exactly that happened in the [Strimzi 0.12 release](https://github.com/strimzi/strimzi-kafka-operator/releases).

Reviewer: The link is to the releases page not to the specific 0.12.1 release.

Author: This was really intentional so that we lead the users to the latest releases rather then one specific. I would probably prefer to keep the link generic.

> And exactly that happened in the [Strimzi 0.12 release](https://github.com/strimzi/strimzi-kafka-operator/releases).

Reviewer 2: ```suggestion
And that's what happened in the [Strimzi 0.12 release](https://github.com/strimzi/strimzi-kafka-operator/releases).
```

> This blog post will tell you more about the improvements to persistent storage in the 0.12 release.

Reviewer 2: ```suggestion
In this post I'm going to tell you more about the improvements to persistent storage in the 0.12 release.
```

<!--more-->

##  Resizing persistent volumes

> Support for resizing volumes depends on the version of your Kubernetes or OpenShift cluster and on the infrastructure it runs on.

Reviewer 2: ```suggestion
Support for resizing volumes depends on the version of your Kubernetes or OpenShift cluster, and on the infrastructure it runs on.
```

We tested this feature with Kubernetes and OpenShift on Amazon AWS.
It should be compatible with Kubernetes 1.11+ and OpenShift 3.11+.
Resizing persistent volumes should work on most major public clouds (such as Amazon AWS, Microsoft Azure and Google Cloud) and many other storage types (such as Cinder or Ceph).

To tell your Kubernetes or OpenShift cluster that your storage supports volume resizing, you have to set the `allowVolumeExpansion` option in your StorageClass to `true`.

> For example following StorageClass would create Amazon AWS GP2 volumes with `xfs` filesystem, encryption and enabled volume expansion:

Reviewer 2: ```suggestion
For example, the following `StorageClass` creates Amazon AWS GP2 volumes with `xfs` filesystem, encryption and volume expansion enabled:
```

```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1beta1
metadata:
  name: ssd
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
provisioner: kubernetes.io/aws-ebs
parameters:
  type: gp2
  encrypted: "true"
  fsType: "xfs"
reclaimPolicy: Delete
allowVolumeExpansion: true
```

The option is called `allowVolumeExpansion` because right now you can only increase the size of the volumes to expand them - you cannot decrease it.

> Strimzi will check that the storage class has this option set to `true` before it tries to resize any volumes.

Reviewer 2: ```suggestion
Strimzi checks that the storage class has this option set to `true` before it tries to resize any volumes.
```

> The [original blog post](https://strimzi.io/2019/02/28/resizing-persistent-volumes.html) explained in detail how does the resizing work:

Reviewer 2: ```suggestion
The [original blog post](https://strimzi.io/2019/02/28/resizing-persistent-volumes.html) explained in detail how this resizing works:
```

1. The size increase is requested by changing the `spec.resources.requests.storage` of the Persistent Volume Claim (PVC)

> 2. Kubernetes will request resizing of the volumes from your infrastructure

Reviewer 2: ```suggestion
2. Kubernetes requests resizing of the volumes from your infrastructure
```

3. Once the resizing of the volume is finished, the pod using this volume needs to be restarted to allow the expansion of the file system

> All what Strimzi does is that it simplifies this process and takes care of it for you.

Reviewer 2: ```suggestion
What Strimzi does is simplify this process by taking care of it for you! 
```

> As a user all you need to do is edit your Kafka custom resource and increase the requested storage size.

Reviewer 2: ```suggestion
All you need to do is edit your Kafka custom resource and increase the requested storage size.
```

You can increase the size of both Zookeeper and Kafka volumes.

> For example you can change the resource from:

Reviewer 2: ```suggestion
For example, you can change the resource from:
```

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
    storage:
      type: persistent-claim
      size: 100Gi
      deleteClaim: false
  zookeeper:
    # ...
    storage:
      type: persistent-claim
      size: 100Gi
      deleteClaim: false
```

to:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
    storage:
      type: persistent-claim
      size: 200Gi
      deleteClaim: false
  zookeeper:
    # ...
    storage:
      type: persistent-claim

>       size: 1000Gi

Reviewer: more realistically do we want to do the opposite increase? 1000 Gi for Kafka and 200 Gi for ZooKeeper, just because in Kafka you store messages so would make sense having bigger size there.

      deleteClaim: false
```

> And the rest will be taken care of by Strimzi.

Reviewer 2: ```suggestion
And the rest is taken care of by Strimzi.
```

> The cluster operator will automatically change the requested volume size in the PVCs and wait until the restart of the pod is required.

Reviewer 2: ```suggestion
The Cluster Operator automatically changes the requested volume size in the PVCs and waits until a restart of the pod is required.
```

> Once the condition of the PVC is set to `FileSystemResizePending`, Strimzi will automatically restart the pod using this PVC.

Reviewer: maybe adding the command line you used for getting the following status information?

Author: The users doesn't need to check the status of the PVCs. The Operator does it. Adding the commands might make it look more complicated than it is. I added a link to the older blog post which has more details about the resizing process.

> Once the condition of the PVC is set to `FileSystemResizePending` (read the [original blog post](https://strimzi.io/2019/02/28/resizing-persistent-volumes.html) for more information about the different states the PVC can be in during the resizing), Strimzi will automatically restart the pod using this PVC.

Reviewer 2: ```suggestion
Once the condition of the PVC is set to `FileSystemResizePending` (read the [original blog post](https://strimzi.io/2019/02/28/resizing-persistent-volumes.html) for more information about the different states the PVC can be in during the resizing), Strimzi automatically restarts the pod using this PVC.
```

```yaml
# ...
status:
  phase: Bound
  accessModes:
    - ReadWriteOnce
  capacity:
    storage: 1000Gi
  conditions:
    - type: FileSystemResizePending
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2019-02-27T12:25:30Z'
      message: >-
        Waiting for user to (re-)start a pod to finish file system resize of
        volume on node.
```

The restart might not happen immediately - it will happen as part of the next periodical reconciliation.

> So you might need to wait couple minutes before it happens.

Reviewer 2: ```suggestion
So you might need to wait a couple of minutes before it happens.
```

> The cluster operator is of course aware of your whole Kafka cluster and will not restart all your pods at the same time but one by one to make sure your cluster will be still in usable state.

Reviewer 2: ```suggestion
The Cluster Operator is, of course, aware of your whole Kafka cluster and will not restart all your pods at the same time, but one by one to make sure your cluster remains in a usable state.
```

## Adding and removing volumes from JBOD storage

JBOD (Just a Bunch Of Disks) storage allows you to use multiple disks in each Kafka broker for storing commit logs.

> Strimzi added support for JBOD storage in Kafka brokers already in version 0.11 (JBOD storage is not supported in Zookeeper).

Reviewer 2: ```suggestion
Strimzi already added support for JBOD storage in Kafka brokers in version 0.11 (JBOD storage is not supported in Zookeeper).
```

> But it didn't allowed you to add or remove volumes from it.

Reviewer 2: ```suggestion
But it didn't allow you to add or remove volumes from it.
```

> You had to specify all your volumes already when creating the cluster.

Reviewer 2: ```suggestion
You had to specify all your volumes when creating the cluster.
```

In 0.12 we significantly improved the JBOD storage support.

> You can of course resize the volumes as described in the previous section.

Reviewer 2: ```suggestion
As well as resizing the volumes as described in the previous section, you can now add or remove volumes in an existing cluster.
```

> But you can now also add or remove volumes in an existing cluster.

Reviewer 2: ```suggestion
```

> To be able to do that, you need to already have an existing Kafka cluster using JBOD storage.

Reviewer 2: ```suggestion
To be able to do that, an existing Kafka cluster already needs to be using JBOD storage.
```

For example:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
    storage:
      type: jbod
      volumes:
      - id: 0
        type: persistent-claim
        size: 1000Gi
        deleteClaim: false
      - id: 1
        type: persistent-claim
        size: 1000Gi
        deleteClaim: false
# ...
```

> If that is what you have, you can easily add more volumes just by editing the YAML.

Reviewer 2: ```suggestion
If you have something like this, you can easily add more volumes just by editing the YAML.
```

For example:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
    storage:
      type: jbod
      volumes:
      - id: 0
        type: persistent-claim
        size: 1000Gi
        deleteClaim: false
      - id: 1
        type: persistent-claim
        size: 1000Gi
        deleteClaim: false
      - id: 10
        type: persistent-claim
        size: 1000Gi
        deleteClaim: false

>       - id: 11

Reviewer: why not id 2 and 3 for the new volumes?

Author: That is to make it more clear that the IDs don't have to be in order.

        type: persistent-claim
        size: 1000Gi
        deleteClaim: false
  # ...
```

Similarly you can also remove volumes:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
    storage:
      type: jbod
      volumes:
      - id: 1
        type: persistent-claim
        size: 1000Gi
        deleteClaim: false
# ...
```

When removing volumes, Strimzi will not do anything to move the data from the volume you are removing.
You have to do that manually before removing the volume(s) from the JBOD storage.
Strimzi will simply remove the volume from the pods.

> **There are some things you should keep in mind and be careful about:**

Reviewer 2: ```suggestion
**There are some things you should keep in mind and take care with:**
```

* The `id` numbers have to be unique.
* The `id` numbers do not have to be in sequence.

> * **Changing the `id` number will be the same as removing the volume with the original `id` and adding a new volume with the new `id`. 

Reviewer 2: ```suggestion
* **Changing the `id` number is the same as removing the volume with the original `id`, and adding a new volume with the new `id`. 
```

Strimzi will stop using the old PVC with the old volume and create a new PVC for the new volume.**
* By default the PVCs are not deleted. 
So if you reuse an `id` which you already used in the past you should check first whether the old PVC still exists or not. 

> If it exists it will be reused instead of creating a new PVC with a new volume.

Reviewer 2: ```suggestion
If it exists, it will be reused instead of creating a new PVC with a new volume.
```

## Using different storage class for each broker

> One of the other storage features we added in 0.12 release is to use different storage class for each broker.

Reviewer 2: ```suggestion
One of the other storage features we added in the 0.12 release is the capability to use a different storage class for each broker.
```

You can specify a different storage class for one or more Kafka brokers, instead of using the same storage class for all of them. 
This is useful if, for example, storage classes are restricted to different availability zones or data centers. You can use the `overrides` field for this purpose.
For example:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
    storage:
      deleteClaim: true
      size: 100Gi
      type: persistent-claim

>       class: my-storage-class

Reviewer: out of curiosity ... if I have a 3 brokers cluster and I want 3 different storage class (one for each broker) as described in this example, what happens if I don't set this "main" `class` field? Does it work? In this scenario, I don't need that because anyway I am going to override classes for each broker in the cluster. Of course, it means that if I increase the replicas and the `class` field isn't here something bad could happen. Just curiosity.

Author: The `class` field is basically the default. If you don't set it, the default storage class will be used as default (i.e. whatever is set as default in Kubernetes). That said, I would not go so deep into this in this post.

      overrides:
        - broker: 0
          class: my-storage-class-zone-1a
        - broker: 1
          class: my-storage-class-zone-1b
        - broker: 2
          class: my-storage-class-zone-1c
  # ...
```

## Conclusion

These improvements should make Strimzi more easy to use.

> They should make it much easier to grow your Kafka cluster together with your project which is important for example to save infrastructure costs.

Reviewer 2: ```suggestion
They should make it much easier to grow your Kafka cluster together with your project, which is important when you are looking to save on infrastructure costs.
```

> Things such as adding volumes or resizing volumes are important part of _Day 2 operations_ which is the area where the operator pattern should be most helpful.

Reviewer 2: ```suggestion
Things such as adding volumes or resizing volumes are an important part of _Day 2 operations_ which is the area where the operator pattern can provide the most help.
```

> In the future Strimzi releases we plan to add more features related to storage - for example backup and recovery.

Reviewer 2: ```suggestion
In future Strimzi releases we plan to add more features related to storage - such as backup and recovery.
```

> If you liked this blog post, star us on [GitHub](https://github.com/strimzi/strimzi-kafka-operator) and follow us on [Twitter](https://twitter.com/strimziio) to make sure you don't miss any of our future blog posts!

Reviewer: Like this idea!
